mesonbuild/modules/pkgconfig.py

```diff
@@ -741,7 +741,15 @@ def parse_variable_list(vardict: T.Dict[str, str]) -> T.List[T.Tuple[str, str]]:
                     self._metadata[lib.get_id()] = MetaData(
                         filebase, name, state.current_node)
         if self.devenv is None:
-            self.devenv = PkgConfigInterface.get_env(state.environment, mesonlib.MachineChoice.HOST, uninstalled=True)
+            pkgbin = state.find_program(
```

Comment:
This is the only place you check for the overriden pkg-config. That means only devenv will have it? I think we need a way that would ensure that meson itself would be using the overriden pkg-config, for consistency.

Reply:
For consistency we also would need to check that override_find_program('pkg-config') happens before we ever use pkg-config earlier. In the same way meson will refuse to override a program we already looked up.

For the implementation, I'm not sure we have a better way than special-case `pkg-config` in Interpreter.add_find_program_override() and give the program to PkgConfigInterface. There we can check if an implementation has already been cached in PkgConfigInterface.class_impl and raise an error.

Reply:
> This is the only place you check for the overriden pkg-config. That means only devenv will have it? I think we need a way that would ensure that meson itself would be using the overriden pkg-config, for consistency.

`PkgConfigInterface` is a singleton that is implicitly initialized when calling `get_env` and it will get called the first time `self.devenv` is `None`. There must be a better way to initialize it.

> For the implementation, I'm not sure we have a better way than special-case pkg-config in Interpreter.add_find_program_override() and give the program to PkgConfigInterface. There we can check if an implementation has already been cached in PkgConfigInterface.class_impl and raise an error.

I think that's a better way to consistently initialize it. I will try this approach.

Reply:
I implemented it following your advice to ensure it's used correctly in all scenarios (my initial implementation was not working for the use case I wanted to fix). I added a new test case that overrides pkg-config and verifies that it's being used correctly. I think it's ready for a final review.